File: implement-shell-tools/wc/my_wc.py
import os
import sys
import argparse

# CLI argument parsing
parser = argparse.ArgumentParser(description="Simplified implementation of wc")
parser.add_argument("paths", nargs="*", default=["."], help="One or more file or directory paths")
parser.add_argument("-l", "--line", action="store_true", help="Count lines")
parser.add_argument("-w", "--word", action="store_true", help="Count words")
parser.add_argument("-c", "--character", action="store_true", help="Count characters")

args = parser.parse_args()

# If no flags are set, show all
show_all = not (args.line or args.word or args.character)

# Count content in a string
def count_content(content):
    lines = content.splitlines()
    words = content.strip().split()
    characters = len(content)
    return len(lines), len(words), characters

# Format output line based on flags
def format_output(lines, words, chars, label):
    parts = []
    if args.line or show_all:
        parts.append(f"{lines:8}")
    if args.word or show_all:
        parts.append(f"{words:8}")
    if args.character or show_all:
        parts.append(f"{chars:8}")
    parts.append(label)
    return " ".join(parts)

# Totals for multiple files
total = {"lines": 0, "words": 0, "characters": 0}
file_count = 0

for input_path in args.paths:
    try:
        if os.path.isdir(input_path):
            print(f"{input_path} is a directory. Skipping.")
            continue

        with open(input_path, "r", encoding="utf-8") as f:
            content = f.read()

        lines, words, characters = count_content(content)

        total["lines"] += lines
        total["words"] += words
        total["characters"] += characters
        file_count += 1

        print(format_output(lines, words, characters, input_path))

    except Exception as e:
        print(f'Error reading "{input_path}": {e}', file=sys.stderr)

# Print totals if more than one file processed
if file_count > 1:
    print(format_output(
        total["lines"],
        total["words"],
        total["characters"],
        "total"
    ))

File: implement-shell-tools/wc/MyWc.js
import { program } from "commander";
import { promises as fs } from "node:fs";

program
  .name("myWc")
  .description("my wc clone")
  .option("-l", "line count")
  .option("-w", "words count")
  .option("-c", "character count")
  .option("-s", "character count without spaces")
  .argument("[paths...]", "file or directory paths");

program.parse();

const opts = program.opts();
let files = program.args;

if (files.length === 0) {
  files = ["."];
}

let totalLines = 0;
let totalWords = 0;

if (opts.l) {
  for (const file of files) {
    const content = await fs.readFile(file, "utf-8");
    const lineCount = content.split("\n").length;

    totalLines += lineCount;
  }
  console.log("Lines:", totalLines);
}
if (opts.w) {
  for (const file of files) {
    const content = await fs.readFile(file, "utf-8");

    const wordCount = content.trim().split(/\s+/).length;
    totalWords += wordCount;
  }
  console.log("Total words:", totalWords);
}
if (opts.c) {
  let totalChars = 0;
  for (const file of files) {
    const content = await fs.readFile(file, "utf-8");

    totalChars += content.trim().length;

  }
  console.log("Total characters:", totalChars);
}

if (opts.s) {
  let totalCharsNoSpaces = 0;
  for (const file of files) {
    const content = await fs.readFile(file, "utf-8");
    const withoutSpaces = content.replace(/\s/g, "");
    totalCharsNoSpaces += withoutSpaces.length;
  }
  console.log("Total characters without spaces:", totalCharsNoSpaces);
}
